### DISCUSSION TITLE
Port 18.04 and Melodic and Semantic segmentation backend

### DISCUSSION THREAD

**Discussion by FPSychotic** [UnknownFreeOccupied/ufomap#3] [Ideas]
Hi, this software looks great, I asked myself if it could support 18.04 Melodic,and if in the future you will add semantic colorization from a NN (with Open3d maybe?) as segnet,detectron2. cuda and ros2 support. 
Some plugin to accept a 3D boundary box (from NN) and fill it as an obstacue to can avoid it and segmented to don't be part of the map, would be amazing.
So easy ask when other need do it...but I build advanced robots as hobby and this is what I miss.
and you did all this alone? holly molly

Cheers!!

**Comment by danielduberg** [comment]
Hi @FPSychotic,

Thank you for your interest and your nice comments!

## UFOMap on Ubuntu 18.04 and ROS Melodic
UFOMap actually works with Ubuntu 18.04 and ROS Melodic as it is now. You only need a compiler that supports C++17 and Intel(R) Threading Building Blocks 2018, which you can install on Ubuntu 18.04. I believe [GNU C++ compiler version 9+ should work](https://github.com/UnknownFreeOccupied/ufomap/issues/2#issue-870500866). Clang does not work since it does not support the [parallel execution policies](https://en.cppreference.com/w/cpp/algorithm/execution_policy_tag), as you can see on the [compiler support page](https://en.cppreference.com/w/cpp/compiler_support) if you search for `std::execution`.

When I used UFOMap on Ubuntu 18.04 I installed the dependencies like this:
### Install GNU C++10 compiler
```bash
sudo add-apt-repository ppa:ubuntu-toolchain-r/test
sudo apt-get update
sudo apt install gcc-10 g++-10
# After this you have to set g++-10 as your compiler
```
### Install TBB 2018+
```bash
echo "deb http://cz.archive.ubuntu.com/ubuntu focal main universe" | sudo tee -a  /etc/apt/sources.list
# or
echo "deb http://cz.archive.ubuntu.com/ubuntu eoan main universe" | sudo tee -a  /etc/apt/sources.list
sudo apt update
sudo apt upgrade
sudo apt install libtbb-dev
```
Be careful with this though, I recommend reading this [Ask Ubuntu answer](https://askubuntu.com/a/1183811).

After this you should be able to [follow the instructions in the Wiki and build UFOMap](https://github.com/UnknownFreeOccupied/ufomap/wiki/Setup).

## Other Questions
Regarding the question about semantics. We have some work on this at the moment, with a first version working. It will hopefully be released relatively soon. I cannot go into too much details about it as of now, as we are investigating and evaluating what kind of functionality and so on we want. However, if I understand what you are saying correctly then I can say that what you want will be supported.

ROS 2 support is on the bucket list. I will add it as soon as I have time. There are some other things that have higher priority right now, but if this is something that is really sought after then I will try to finish it faster. I continually make improvements and introduce new features and fixes to UFOMap, so I am very happy to take suggestions to help you be able to build what you want :smiley:

Thank you a lot for your interest. I am very interested in seeing what you manage to create with UFOMap, please share if you like :smile:

Cheers!

**Reply by generush** [reply]
Hello @danielduberg, I'm trying to get ufomap installed on my Ubuntu 18.04 computer so I can play around with in ROS, and have been following these instructions. I got through the instructions, but am currently stuck. Specifically, I wanted to share two clarifying points that might help other folks with this setup process, and ask one question about a build error that I am facing:

_Clarifying Point 1_ 

After installing gcc-10 and g++-10, I had to switch to it using:
`sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-10 10`.
I found instructions [here](https://linuxconfig.org/how-to-switch-between-multiple-gcc-and-g-compiler-versions-on-ubuntu-20-04-lts-focal-fossa) for doing so.

_Clarifying Point 2_

Additionally, eoan did not allow me to update TBB from my 2017 version to 2020
`'echo "deb http://cz.archive.ubuntu.com/ubuntu eoan main universe" | sudo tee -a  /etc/apt/sources.list'`
 but focal did:
`'echo "deb http://cz.archive.ubuntu.com/ubuntu focal main universe" | sudo tee -a  /etc/apt/sources.list'.`

_**Question**_

When I try building, I receive an error `install TARGETS given no LIBRARY DESTINATION for shared library target
  "Map".` Do you happen to know how to resolve this error? Seems like there is a library called Map and it's not finding that library to link against for some reason?

![image](https://user-images.githubusercontent.com/33467025/117486192-92b6be80-af26-11eb-9a35-67c2c60059a2.png)
[error.txt](https://github.com/UnknownFreeOccupied/ufomap/files/6443285/error.txt)

Thanks in advance for the help!

**Reply by FPSychotic** [reply]
I arrived more or less to the same point, but I was able to install TBB from a PPA, but got broken by other side. Probably you are in arm64 like me as it doesnt allow you update tbb, if you are in x86 I will try in a X86 one of this days, I will let you know any advance I make

**Reply by generush** [reply]
Hello,

To clarify, I was able to update TBB from PPA. 

My build issue is now resolved, a colleague of mine helped me fix the library link error by replacing the following line 150 in the CMakeLists.txt file in ufomap:
`
	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}`
with the new line
`
	LIBRARY DESTINATION lib ${CMAKE_INSTALL_LIBDIR}`.

I am now batting Qt. I was able to build all packages except the [ufomap_rviz_plugins](url).
![image](https://user-images.githubusercontent.com/33467025/117501730-b20c1680-af3b-11eb-8b52-0f78c2348287.png)

I believe I need to install Qt 5.10+ on my Ubuntu 18.04, I currently have 5.9.5.

@FPSychotic did you have this Qt issue and were you able to figure it out on Ubuntu 18.04?

**Reply by FPSychotic** [reply]
@erush91 No, I couldn't arrive there, I got stuck before. Are you in Arm64 or in X86?, if you are in ARM64 I can try help you tomorrow after try your solution, if you are in x86 probably we are in different issue

**Reply by generush** [reply]
@danielduberg We found an easy workaround which was to comment out the few lines of code that required the QLocale library. It didn't look critical to the mapping functionality. Now it works with Qt 5.9.5 and I don't need to upgrade to 5.10+ (which was proving to be very difficult on ROS Melodic).

Thanks!

**Reply by FPSychotic** [reply]
@erush91 I'm facing similar issues how you did fix the QT issue and in what platform you are, x86 or arm64?that will help thanks

**Reply by generush** [reply]
I'm using x86. My computer is an [Intel NUC](https://simplynuc.com/7i7dnbe-shop/), so perhaps our problems differ.

Rather than installing Qt 5.10+ to resolve that compiler error, I simply commented out the references to the QLocale library, since it seems it's not really needed for the core functionality:

https://github.com/UnknownFreeOccupied/ufomap/blob/5b0a45d5c3867d4fd1747ebb1d9aa26185e7656f/ufomap_ros/ufomap_rviz_plugins/src/ufomap_display.cpp#L46

https://github.com/UnknownFreeOccupied/ufomap/blob/5b0a45d5c3867d4fd1747ebb1d9aa26185e7656f/ufomap_ros/ufomap_rviz_plugins/src/ufomap_display.cpp#L585

**Comment by FPSychotic** [comment]
Thanks so much by the great explanation, the good news about future features and your fast and kind answer.
I will follow the project with interest.
Of course I will share experiences, new suggestions and show whatever I build with your project inside.
I have 3 rovers with different setups and a drone that could receive your software, included future ros2 versions, ML features as cuda, tpu, opencl, or different sensors as d435,d455,t265,r200,rpi cameras,fisheye cameras. They are based in jetson nano, xavier Nx, Rpi4 and Nuc x86 and Up x86, as AI acceleration for non nVidia SoC I have TPU and K210.
2 of them are Ardupilot+ROS and others full ROS

Apart that I will share experiences with them, let me know if you want I test something on them, whatever can help you, I made them for R&D but I cannot use them as I'm not programmer ,developer and I have not technical background, I just can add what others make and I make them with many effort because I need supply the lack of knowledge.
But I think they are quite nice, and already autonomous. 
They are at your (and your  team) full disposition, what better use they could have,? they are for that. I cannot use them properly as I have not technical background and is a little sad.

If you allow me I show you some pictures.
![i-Hexa atom-1](https://user-images.githubusercontent.com/36112212/117053599-113a1300-ad11-11eb-92c9-22e5d0002a6c.png)
![i-nROVER 200D (2)](https://user-images.githubusercontent.com/36112212/117053604-11d2a980-ad11-11eb-85e7-ba1e9e084f53.png)

![IMG_20210417_021810](https://user-images.githubusercontent.com/36112212/117051870-ecdd3700-ad0e-11eb-871d-33e3e2be488b.jpg)
![IMG_20210403_115744](https://user-images.githubusercontent.com/36112212/117051883-ee0e6400-ad0e-11eb-88e2-0b49f2691eb4.jpg)
![IMG_20210430_154710](https://user-images.githubusercontent.com/36112212/117051886-ef3f9100-ad0e-11eb-8bad-7311b4a52c6c.jpg)

I'm building the big one now.  The drone has not cameras because they are fitted in the small rover.
Two or three  examples where your software could fit :
https://youtu.be/Aa-6jMenjCY
https://youtu.be/PeqdgwcGpjk
https://youtu.be/1fcpkMFboTQ
https://youtu.be/TTvQactokCU

I have many ideas for real uses for software and rovers /drones like that but I dont know program and I dont have skill  with cad or knowledge to do something commercial, so they are just for hobby and I will try learn program with them, but my brain is rusted 45years and too much Street Fighter gaming in my back XD.

**Comment by FPSychotic** [comment]
I'm trying install it with your instructions but I'm in arm64 (jetson xavier nx) but there is not tbb 2008+. I only found one in ppa but it will brake the ros2 and ros2 instalation.
in my drone there is ubuntu 20.04 (not sure LoL) , but even if not, it is x86, so I will try there. Is more is the ideal place as it has less resources and use intensively optomap
i will update here, maybe helps someone.
PPA with libtbb 2019 for arm64
sudo apt-add-repository ppa:nnstreamer
install it from synaptic
 I guess it is possible compile ros with it. Probably I will in try it soon, but now I have other faster way.
